bgpv1: Add `cilium-dbg bgp route-policies` command & include it in the bugtool

rastislavs opened the pull request:
Adds a new `/bgp/route-policies` API endpoint and the `cilium-dbg bgp route-policies` command that can be used for retrieval of BGP route policies configured in the BGP Control Plane's Router implementation. Also adds this command's output in the bugtool.

This is a debug-ability enhancement for the CiliumBGPPeeringPolicy's `AdvertisedPathAttributes` feature added by https://github.com/cilium/cilium/pull/27705
The command prints the route-policies generated  internally by the BGP Control Plane - these are not directly configurable by the user.

**Example usage:**

```
$ # cilium bgp route-policies
VRouter   Policy Name                                   Type     Match Peers    Match Prefixes (Min..Max Len)   RIB Action   Path Actions
65001     172.0.0.1/32-CiliumLoadBalancerIPPool-32...   export   172.0.0.1/32   192.168.100.0/24 (32..32)       none         AddCommunities: [65001:100]                                       
65001     172.0.0.1/32-PodCIDR                          export   172.0.0.1/32   10.244.0.0/24 (24..24)          none         {SetLocalPreference: 150} {AddCommunities: [56564:1 65001:150]} 
```

```yaml
$ cilium-dbg bgp route-policies -o yaml
- name: 172.0.0.1/32-CiliumLoadBalancerIPPool-325f8fc93e320b4bb1b1ea186c0f2d53e787175a3b5725c4be96cfd93fcf8f2e
  routerasn: 65001
  statements:
    - addcommunities:
        - 65001:100
      addlargecommunities: []
      matchneighbors:
        - 172.0.0.1/32
      matchprefixes:
        - cidr: 192.168.10.128/25
          prefixlenmax: 32
          prefixlenmin: 32
      routeaction: none
      setlocalpreference: -1
    - addcommunities:
        - 65001:100
      addlargecommunities: []
      matchneighbors:
        - 172.0.0.1/32
      matchprefixes:
        - cidr: 2004::/64
          prefixlenmax: 128
          prefixlenmin: 128
      routeaction: none
      setlocalpreference: -1
  type: export
- name: 172.0.0.1/32-PodCIDR
  routerasn: 65001
  statements:
    - addcommunities:
        - "56564:1"
        - 65001:150
      addlargecommunities: []
      matchneighbors:
        - 172.0.0.1/32
      matchprefixes:
        - cidr: 10.244.0.0/24
          prefixlenmax: 24
          prefixlenmin: 24
      routeaction: none
      setlocalpreference: 150
  type: export
```

**Config used for this example:**

Excerpt from `CiliumBGPPeeringPolicy`:
```yaml
apiVersion: cilium.io/v2alpha1
kind: CiliumBGPPeeringPolicy
  name: tor
spec:
  virtualRouters:
  - exportPodCIDR: true
    localASN: 65001
    neighbors:
    - peerASN: 65000
      peerAddress: 172.0.0.1/32
      advertisedPathAttributes:
      - selectorType: CiliumLoadBalancerIPPool
        selector:
          matchLabels:
            environment: production
        communities:
          standard:
          - 65001:100
      - selectorType: PodCIDR
        communities:
          standard:
          - 65001:150
          - "56564:1"
        localPreference: 150
```

LB Pool:
```yaml
apiVersion: cilium.io/v2alpha1
kind: CiliumLoadBalancerIPPool
metadata:
  labels:
    environment: production
  name: cilium-pool
spec:
  cidrs:
  - cidr: 192.168.10.128/25
  - cidr: 2004::0/64
```

rastislavs committed:
\api: Add a new API endpoint `/bgp/route-policies`

Add a new API endpoint `/bgp/route-policies` to retrieve route
policies configured in the BGP Control Plane.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

rastislavs committed:
api,bgpv1: Implement `/bgp/route-policies` API in BGPRouterManager

Implement `/bgp/route-policies` API by wiring-up API call with
BGPRouterManager and GetRoutePolicies API of the Router implementation.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

rastislavs committed:
cilium-dbg: Implement `bgp route-policies` command

Implement `cilium-dbg bgp route-policies` command that retrieves
route policies from the `/bgp/route-policies` API and displays the response.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

rastislavs committed:
bgpv1: Add GetRoutePolicies Router method and implement it in GoBGP

Adds a new GetRoutePolicies() method for retrieving configured route policies
to the Router interface and implements it in the GoBGP Router implementation.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

qmonnet reviewed (approved):
Doc changes (are minimal and) look good. I only gave a quick look at the implementation, didn't look at the details.
Thanks!

harsimran-pabla commented on cilium-dbg/cmd/bgp_route_policy_get.go:
```diff
@@ -0,0 +1,131 @@
+	for _, policy := range policies {
+		fmt.Fprintf(w, "%d\t", policy.RouterAsn)
+		fmt.Fprintf(w, "%s\t", trimString(policy.Name, 40))
+		fmt.Fprintf(w, "%s\t", policy.Type)
```
if len(policy.Statements) = 0, output will not contain new line after printing 3 items.

harsimran-pabla commented on cilium-dbg/cmd/bgp_route_policy_get.go:
```diff
@@ -0,0 +1,131 @@
+		return policies[i].RouterAsn < policies[j].RouterAsn && policies[i].Name < policies[j].Name
+	})
+
+	fmt.Fprintln(w, "VRouter\tPolicy Name\tType\tMatch Peers\tMatch Prefixes\tRIB Action\tPath Actions")
```
```suggestion
	fmt.Fprintln(w, "VRouter\tPolicy Name\tType\tMatch Peers\tMatch Prefixes (Min..Max Len)\tRIB Action\tPath Actions")
```

harsimran-pabla commented on pkg/bgpv1/api/conversions.go:
```diff
@@ -261,3 +261,192 @@ func ToAgentRoutes(ms []*models.BgpRoute) ([]*types.Route, error) {
+		})
+	}
+
+	if len(errs) != 0 {
```
nit: I think this pattern of error handling can be simplified. I don't have strong opinion about it, so feel free to ignore this one.

```
var err error

for _, something := somethings {
  _, sErr := func(something)
  if sErr != nil {
    err = errors.Join(err, sErr)
  }

return err
}

or 

return errors.Join(errs...)
```

I think errors.Join will return non-nil err only if one of the errors in the list is non-nil.

harsimran-pabla commented on pkg/bgpv1/types/test_fixtures.go:
```diff
@@ -40,4 +43,131 @@ var (
 	}
+
+	// CommonRoutePolicies contains common route policy values appearing in the agent code
+	CommonRoutePolicies = []struct {
```
nit: consider renaming it, for eg `TestCommonRoutePolicies`. Reference from other parts of the code like types.CommonRoutePolicies is not explicit that it is test code.

harsimran-pabla commented on pkg/bgpv1/gobgp/state.go:
```diff
@@ -281,3 +281,49 @@ func (g *GoBGPServer) GetRoutes(ctx context.Context, r *types.GetRoutesRequest)
+
+	// list policy assignments into a map for later use
+	assignments := make(map[string]*gobgp.PolicyAssignment)
+	err = g.server.ListPolicyAssignment(context.Background(), &gobgp.ListPolicyAssignmentRequest{}, func(a *gobgp.PolicyAssignment) {
```
```suggestion
	err = g.server.ListPolicyAssignment(ctx, &gobgp.ListPolicyAssignmentRequest{}, func(a *gobgp.PolicyAssignment) {
```

harsimran-pabla commented on pkg/bgpv1/gobgp/state.go:
```diff
@@ -281,3 +281,49 @@ func (g *GoBGPServer) GetRoutes(ctx context.Context, r *types.GetRoutesRequest)
+	// list & convert policies
+	var policies []*types.RoutePolicy
+	err = g.server.ListPolicy(ctx, &gobgp.ListPolicyRequest{}, func(p *gobgp.Policy) {
+		assignment := assignments[p.Name]
```
nit : use map lookup boolean field for existence check.
```
assignment, exists := assignments[p.Name]
if exists {
  ...
}
```

assignment is point type, but if someday we change its type to non-pointer, assignment != nil will always be true.

rastislavs replied:
fixed in https://github.com/cilium/cilium/pull/28973/commits/b88fa8491d5e4335417dac7a3d216039b7ca8c43

harsimran-pabla reviewed (changes requested):
change looks good, some minor comments.

rastislavs committed:
bugtool: Include `cilium-dbg bgp route-policies` command output

Includes BGP route policies output in the bugtool to have this
BGP Control Plane information included in the sysdumps.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

harsimran-pabla reviewed (approved):
thanks, lgtm!

derailed commented on daemon/cmd/bgp.go:
```diff
@@ -39,3 +39,17 @@ func getBGPRoutesHandler(d *Daemon, params restapi.GetBgpRoutesParams) middlewar
+// getBGPRoutePoliciesHandler gets BGP routing policies from BGP controller
+func getBGPRoutePoliciesHandler(d *Daemon, params restapi.GetBgpRoutePoliciesParams) middleware.Responder {
+	if d.bgpControlPlaneController == nil {
+		return api.Error(http.StatusNotImplemented, fmt.Errorf("BGP Control Plane disabled"))
```
nit: errors.New("xxx") no caps

derailed commented on pkg/bgpv1/api/conversions.go:
```diff
@@ -261,3 +261,180 @@ func ToAgentRoutes(ms []*models.BgpRoute) ([]*types.Route, error) {
 }
+
+func ToAPIRoutePolicies(policies []*types.RoutePolicy, routerASN int64) []*models.BgpRoutePolicy {
+	var ret []*models.BgpRoutePolicy
```
Might make sense to pre-alloc the slices here and below if expecting many entries??

derailed commented on pkg/bgpv1/api/conversions.go:
```diff
@@ -261,3 +261,180 @@ func ToAgentRoutes(ms []*models.BgpRoute) ([]*types.Route, error) {
+
+func ToApiRoutePolicyType(t types.RoutePolicyType) string {
+	if t == types.RoutePolicyTypeExport {
+		return "export"
```
Since we are checking on these strings might make sense to make them constants?

derailed commented on pkg/bgpv1/api/conversions.go:
```diff
@@ -261,3 +261,180 @@ func ToAgentRoutes(ms []*models.BgpRoute) ([]*types.Route, error) {
+func ToApiRoutePolicyAction(a types.RoutePolicyAction) string {
+	switch a {
+	case types.RoutePolicyActionNone:
+		return "none"
```
Since we are checking on these strings might make sense to make them constants?

rastislavs replied:
fixed (and the other comments as well) in https://github.com/cilium/cilium/pull/28973/commits/4360104ce89629213e1c9e77a9330f1bed2c7904

derailed reviewed (commented):
@rastislavs Nice work!

derailed reviewed (approved):
@rastislavs LG. Thanks for the updates!

nathanjsweet reviewed (approved):
LGTM for API
